src/main/java/io/netty/incubator/codec/quic/QuicheQuicChannel.java

```diff
@@ -782,53 +782,56 @@ void writable() {
-                    if (capacity == 0) {
-                        // Still not writable, put back in the queue.
-                        flushPendingQueue.add(streamId);
-                    } else {
-                        long sid = streamId;
-                        QuicheQuicStreamChannel channel = streams.get(sid);
-                        if (channel != null) {
-                            if (capacity > 0) {
-                                mayNeedWrite = true;
-                                channel.writable(capacity);
-                            } else {
-                                if (!Quiche.quiche_conn_stream_finished(connAddr, sid)) {
-                                    // Only fire an exception if the error was not caused because the stream is
-                                    // considered finished.
-                                    channel.pipeline().fireExceptionCaught(Quiche.newException(capacity));
+                long writableIterator = Quiche.quiche_conn_writable(connAddr);
+
+                try {
+                    // For streams we always process all streams when at least on read was requested.
+                    for (;;) {
+                        int writable = Quiche.quiche_stream_iter_next(
+                                writableIterator, writableStreams);
+                        for (int i = 0; i < writable; i++) {
+                            long streamId = writableStreams[i];
+                            QuicheQuicStreamChannel streamChannel = streams.get(streamId);
+                            if (streamChannel != null) {
+                                int capacity = Quiche.quiche_conn_stream_capacity(connAddr, streamId);
+                                if (capacity < 0) {
+                                    if (!Quiche.quiche_conn_stream_finished(connAddr, streamId)) {
+                                        // Only fire an exception if the error was not caused because the stream is
+                                        // considered finished.
+                                        streamChannel.pipeline().fireExceptionCaught(Quiche.newException(capacity));
+                                    }
+                                    // Let's close the channel if quiche_conn_stream_capacity(...) returns an error.
+                                    streamChannel.forceClose();
+                                } else if (streamChannel.writable(capacity)) {
+                                    mayNeedWrite = true;
                                 }
-                                // Let's close the channel if quiche_conn_stream_capacity(...) returns an error.
-                                channel.forceClose();
                             }
                         }
+                        if (writable < writableStreams.length) {
+                            // We did handle all writable streams.
+                            break;
```

Comment:
This looks weird. We break out of the `writableIterator` loop if the `writable` is less than `writableStreams.length`.

If it is possible for `writable` to be greater than `writableStreams.length`, then we would get `ArrayIndexOutOfBoundsException` from the for-loop above.

So something isn't right with the logic here.

Reply:
@chrisvest nope... so how this works is basically that you pass the iterator pointer and an array to the native method. We then fill the array and return the number of elements we put in the array. Which means if the number of elements (writable) is less then the length we did reach the end of the iterator.

Reply:
Aha, the writableStreams is an out-parameter there. Makes sense.